### PR TITLE
Improve types and error support in Logger.error

### PR DESCRIPTION
**Pull request by samueljseay**
### Changes proposed in this Pull Request:

In https://github.com/woocommerce/woocommerce/pull/38138 I went down a rabbit hole trying to improve a number of typing problems in CLI core, but it felt like too much refactoring for too little value. But I did feel like one valuable thing came out of it which was to improve the `Logger` so this PR introduces just those changes.

in JS anything can be thrown, and so TS correctly types it as `unknown`. But that means you often have to check the error is `instanceof Error` or similar. 

That makes this current pattern in our code not work well with TS:

```
} catch ( e ) {
  Logger.error( e );
 }
 ```
 And also from a runtime perspective that may not always log something useful. This PR allows `Logger.error` to be passed `unknown` and does a best effort based on error type to log something.
 
 I also introduced a backwards compatible flag for turning off process.exit if you want to log but don't want the program to exit. I think you could argue I should not have made exit the responsibility of the Logger, it just felt convenient when we started writing these scripts to exit from the Logger since it's outputting a fatal error. At least now you can choose.

### How to test the changes in this Pull Request:

Tests were introduced that cover the original functionality and new functionality, so I don't think we need to test here beyond that. Before if you logged a non-error type it would just output worse output for the error, at least now it will JSON stringify anything that's not an error and thats better than nothing!

### CODE REVIEW

File: tools/monorepo-utils/src/core/logger.ts [hunk 1]
```diff
@@ -25,10 +25,20 @@ export class Logger {
...
 		if ( Logger.loggingLevel >= LOGGING_LEVELS.error ) {
-			error( chalk.red( message ) );
-			process.exit( 1 );
+			if ( err instanceof Error ) {
+				error( chalk.red( err.message ) );
+			} else if ( typeof err === 'string' ) {
+				error( chalk.red( err ) );
+			} else {
+				// Best effort to log the error when we don't know the type.
+				error( chalk.red( JSON.stringify( err ) ) );
```
Comment (psealock) [inline tools/monorepo-utils/src/core/logger.ts]: If we know the error is an object and you want to stringify it, then passing in these extra arguments makes it readable:

```
console.log(JSON.stringify({foo:1, bar:2}, null, 4));
```
See https://stackoverflow.com/questions/18184559/how-to-pretty-print-log-output-in-chrome-devtools-console
And you can check in the console

File: tools/monorepo-utils/src/core/logger.ts [hunk 2]
```diff
@@ -25,10 +25,20 @@ export class Logger {
 		] as number;
 	}
 
-	static error( message: string ) {
+	static error( err: unknown, failOnErr = true ) {
 		if ( Logger.loggingLevel >= LOGGING_LEVELS.error ) {
-			error( chalk.red( message ) );
-			process.exit( 1 );
+			if ( err instanceof Error ) {
+				error( chalk.red( err.message ) );
```
Comment (psealock) [inline tools/monorepo-utils/src/core/logger.ts:31]: Nice!

**Review by psealock — COMMENTED** [review]
Nice changes here. Just a suggestion about readability for objects

File: tools/monorepo-utils/src/core/logger.ts (hunk 1 above)
Comment (samueljseay) [reply tools/monorepo-utils/src/core/logger.ts]: Oh yeah the indentation looks better that way, thanks!

Comment (samueljseay) [reply tools/monorepo-utils/src/core/logger.ts]: @psealock resolved in https://github.com/woocommerce/woocommerce/pull/38160/commits/4068828a08e924391db16d9079cd4deb2d1dcd0b

Note I just went with 2 spaces just to keep the output tighter for large objects. It still looks more readable. Thanks for the suggestion.

**Review by psealock — APPROVED** [review]
🚀

**Review by psealock — APPROVED** [review]
🚀

**Pull request merged**